From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samuel Attard <sattard@anthropic.com>
Date: Fri, 15 Aug 2025 14:58:12 -0700
Subject: feat: add support for embedder snapshot validation

IsValid is not exposed despite being commented as for embedders, this exposes something that works for us.

diff --git a/gin/v8_initializer.cc b/gin/v8_initializer.cc
index 1a6063067108b35b8bd71c02727618a720b06c0e..6ac0c4d180967c14ca7e41c08d6ca4d0019096e6 100644
--- a/gin/v8_initializer.cc
+++ b/gin/v8_initializer.cc
@@ -80,11 +80,23 @@ bool GenerateEntropy(unsigned char* buffer, size_t amount) {
   return true;
 }
 
+static base::RepeatingCallback<void(v8::StartupData*)>& SnapshotValidator() {
+  static base::NoDestructor<base::RepeatingCallback<void(v8::StartupData*)>>
+      validator(
+          base::BindRepeating([](v8::StartupData* data) -> void { /* empty */ }));
+  return *validator;
+}
+
+void SetV8SnapshotValidatorInner(const base::RepeatingCallback<void(v8::StartupData*)>& callback) {
+  SnapshotValidator() = std::move(callback);
+}
+
 void GetMappedFileData(base::MemoryMappedFile* mapped_file,
                        v8::StartupData* data) {
   if (mapped_file) {
     data->data = reinterpret_cast<const char*>(mapped_file->bytes().data());
     data->raw_size = static_cast<int>(mapped_file->bytes().size());
+    SnapshotValidator().Run(data);
   } else {
     data->data = nullptr;
     data->raw_size = 0;
@@ -250,6 +262,10 @@ constexpr std::string_view kV8FlagParam = "V8FlagParam";
 
 }  // namespace
 
+void SetV8SnapshotValidator(const base::RepeatingCallback<void(v8::StartupData*)>& callback) {
+  SetV8SnapshotValidatorInner(std::move(callback));
+}
+
 class V8FeatureVisitor : public base::FeatureVisitor {
  public:
   void Visit(const std::string& feature_name,
diff --git a/gin/v8_initializer.h b/gin/v8_initializer.h
index c880e5bec57882df2317e7a61108755b3377f173..810a6a29b85791968ab8a037b438895e7d9e235b 100644
--- a/gin/v8_initializer.h
+++ b/gin/v8_initializer.h
@@ -10,6 +10,7 @@
 #include "base/files/file.h"
 #include "base/files/file_path.h"
 #include "base/files/memory_mapped_file.h"
+#include "base/functional/callback.h"
 #include "build/build_config.h"
 #include "gin/array_buffer.h"
 #include "gin/gin_export.h"
@@ -27,6 +28,8 @@ class StartupData;
 
 namespace gin {
 
+void SetV8SnapshotValidator(const base::RepeatingCallback<void(v8::StartupData*)>& callback);
+
 class GIN_EXPORT V8Initializer {
  public:
   // This should be called by IsolateHolder::Initialize().
